source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -82,71 +96,83 @@ void AuthenticatorImpl::sanitizePayloadHeaders(Http::HeaderMap& headers) const {
     }
   }
 }
-
 void AuthenticatorImpl::verify(Http::HeaderMap& headers, Authenticator::Callbacks* callback) {
   headers_ = &headers;
   callback_ = callback;
 
   ENVOY_LOG(debug, "Jwt authentication starts");
-  auto tokens = config_->getExtractor().extract(headers);
-  if (tokens.empty()) {
-    if (okToBypass()) {
-      doneWithStatus(Status::Ok);
-    } else {
-      doneWithStatus(Status::JwtMissed);
-    }
+  if (!issuer_opt_) {
+    tokens_ = config_->getExtractor().extract(headers);
+  }
+  if (tokens_.empty()) {
+    doneWithStatus(Status::JwtMissed);
     return;
   }
 
-  // TODO(qiwzhang), add supports for multiple tokens.
-  // Only process the first token for now.
-  token_.swap(tokens[0]);
+  startVerify();
+}
 
-  const Status status = jwt_.parseFromString(token_->token());
-  if (status != Status::Ok) {
-    doneWithStatus(status);
-    return;
-  }
+void AuthenticatorImpl::startVerify() {
+  Status status;
+  while (!tokens_.empty()) {
+    jwt_ = {};
+    status = jwt_.parseFromString(tokens_.back()->token());
```

Comment:
My suggestion is: use another member variable  curr_token_ to store the current working token and call pop_back() here.  so no need to have so many places with back() and  pop_back() calls.